IA-3762 useUniqueIdFn helper hook

msilva-broad opened the pull request:
- part of work for IA team conversion to TS and Unit tests push (Export Anaysis Modal).
- promoted IdContainer.js to TS file.
- marked IDContainer as DEPRECATED.  The pattern it promotes is a bit convoluted in its functional nesting, and creates Typescript headaches.
- implemented and unit tested alternative useUniqueIdFn helper hook pattern. See jsdocs and unit tests for details.

---
- based on PR feedback, simplified hook to be useUniqueId and return the id directly instead of a callable function.
- useUniqueId takes optional prefix at hook creation time, will default to 'element' to match previous behavior.
- see updated jsdocs and unit tests.

msilva-broad committed:
IA-3762 useUniqueIdFn helper hook
- part of work for IA team conversion to TS and Unit tests push (Export Anaysis Modal).
- promoted IdContainer.js to TS file.
- marked IDContainer as DEPRECATED.  The pattern it promotes is a bit convoluted in its functional nesting, and creates Typescript headaches.
- implemented and unit tested alternative useUniqueIdFn helper hook pattern. See jsdocs and unit tests for details.

nawatts commented on src/components/common/IdContainer.ts:
```diff
@@ -0,0 +1,32 @@
+ */
+export const useUniqueIdFn = (): UniqueIdFn => {
+  const uniqueId: string = useMemo(() => _.uniqueId(''), [])
+  return ((prefix: string): string => {
+    return `${prefix}-${uniqueId}`
+  })
```
What's the purpose of having the hook return a function to get the ID instead of returning the ID itself?

From the test, I can guess the intent of having it return a function:
```js
const getMyId = useUniqueIdFn()

const buttonAId = getMyId('button-a')
const buttonBId = getMyId('button-b')
```

However, it seems like that pattern could easily be achieved with:
```js
const myId = useUniqueId()

const buttonAId = `button-a-${myId}`
const buttonBId = `button-b-${myId}`
```

msilva-broad replied:
I had the same quesiton/debate in my head.... 
- getMyId fn is a little easier on they keystrokes (no ` and $ keys required)
- but the `prefix-${myId}` pattern is more transparent and simpler on the return...

The tie-breaker for me was to more closely model the pattern this was replacing... which worked the prefix into the flow of things... but make the prefix authorable.

I think that, and unique('button-a') seemed to read nicely to me.

nawatts replied:
I think the most common use case for IdContainer is linking form elements and their labels.

I think I'd rather write:
```js
const id = useUniqueId()

...
label({ htmlFor: id }, ...),
h(Input, { id }),
```

Than
```js
const getId = useUniqueIdFn()
const id = getId()

...
label({ htmlFor: id }, ...),
h(Input, { id }),
```

or

```js
const getId = useUniqueIdFn()

...
label({ htmlFor: getId() }, ...),
h(Input, { id: getId() }),
```
 
And having the hook return an ID instead of a function leaves it up to the consumer how/if they want to construct derived IDs, like adding a prefix or a suffix.

cahrens replied:
I strongly prefer the 1st of these 3 options. I would expect every time that you call `useUniqueId()`, you get a unique ID. Otherwise I really don't understand the point of the method.

But again, what I really want to do most of the time is simply nest the input inside of the label, as then you don't need the ids at all. Is there a way to do this with our way of writing React?

```
<label>
  First name:
  <input type="text" name="firstname">
</label>

nawatts replied:
> Is there a way to do this with our way of writing React?

Yes.

```js
label([
  'First name:',
  h(TextInput, { name: 'firstname' })
])
```

cahrens replied:
Ah right, that looked odd to me but obviously….

I would suggest we encourage people to write more in this style, as then ids are needed far less often. Assuming of course that we don't emit a false alarm in this case via the `id` assertion.

nawatts commented on src/components/common/IdContainer.ts:
```diff
@@ -0,0 +1,32 @@
+ * // my-thing-two-123UniqueToken123
+ */
+export const useUniqueIdFn = (): UniqueIdFn => {
+  const uniqueId: string = useMemo(() => _.uniqueId(''), [])
```
IdContainer uses `useState` while this uses `useMemo`. Both make sense to me, but `useState` guarantees that the unique ID remains the same throughout the component's lifetime. Should both IdContainer and useUniqueIdFn have the same behavior?

msilva-broad replied:
my understanding is that useMemo is more optimized to cache things that are not expected to change, which is why I chose it.

nawatts replied:
Yeah, `useMemo` makes sense. I was wondering if we wanted to preserve IdContainer's guarantee that the ID doesn't change?

nawatts replied:
Then again, I think I've convinced myself there's no need for the additional guarantee and that useMemo is fine here.

nawatts reviewed (commented):
> marked IDContainer as DEPRECATED. The pattern it promotes is a bit convoluted in its functional nesting, and creates Typescript headaches.

Agreed that a hook is easier to use than an HOC here. I think the TypeScript headaches would be solved by typing IdContainer's `children` prop so that it's not assumed to be a React node.

nawatts commented:
Also, question on organization... do we want this hook in the IdContainer module or its own module?

msilva-broad commented:
> 

yes, but I also want to bias away from deeply nested code patterns, which was raised as a contributor to making the code harder to absorb/work-with.

msilva-broad commented:
> Also, question on organization... do we want this hook in the IdContainer module or its own module?

Probably the better outcome would be to rename the file to uniqueId.ts.  But I didn't want to do that at the same time as promoting to Typescript and also add PR noise for import reference changes.

cahrens commented on src/components/common/IdContainer.ts:
```diff
@@ -0,0 +1,32 @@
+ * const unique = useUniqueIdFn()
+ * //...
+ * div({ id: unique('my-thing-one') }, [...],
+ * div({ id: unique('my-thing-two') }, [...]
```
In general one only needs an id for linking together labels/inputs (which actually could be more easily done simply by nesting the input inside of the label component) for linking for things like aria-describedby. Can you make the example show a pattern of how it would actually be used?

cahrens replied:
Also, is the prefix optional? I don't really understand why a prefix is needed at all, so there should be a sensible default value.

msilva-broad replied:
right now it is not optional... see my other comment.  But I am happy for this conversation :)

cahrens commented on src/components/common/IDContainer.test.ts:
```diff
@@ -0,0 +1,41 @@
+    const otherId = getMyId2('other')
+
+    // Assert
+    // id can be reused within a component and will be same seed
```
Why would you ever want to reuse the id? I think of the point of this function just that it gives you an id that is guaranteed to be unique on the page.

msilva-broad replied:
Right now the code is spitting out non-human-readable ids.  Everything is prefixed with 'element-123uniqueToken123'.  The intention was to make things a bit more human readable.  The ID would be intended to be re-used, and sub-components would tag on different prefixes... hence the examples of button-a-123, button-b-123. 

The proposed implementation has the hook return a callable function, which just prefixes a cached unique ID for the component which survives multiple render cycles.  This would have the benefit of also helping debugging since the ID tokens would match for a given component's sub-components.

But as was stated in other comments, this should be an uncommon need that we can largely remove with tweaks/additions to our common component offerings and/or recommended patterns for things like form labels, tab controls, etc.

I do see the benefit of @nawatts suggestion to just have the hook return the raw id.  That makes the useage more transparent.... but I want to encourage the button-a-123 pattern, and just returning '123' doesn't promote that very well...

msilva-broad replied:
also, the intent was to primarily protect against duplicated instances of a control within the application... so that instanceA of MyCoolComponent and instanceB elsewhere in the app can have unique Ids from eachother.   The specific proposed solution is to solve this by also making the Ids more human-readable and debug-friendly.

cahrens replied:
_"The ID would be intended to be re-used, and sub-components would tag on different prefixes... hence the examples of button-a-123, button-b-123."_

I am really not convinced this was the original intention, and it has the potential of folks creating non-unique IDs, which defeats the purpose of having the utility method. There is really no reason that an ID needs to be human-readable-- it is purely used to associate elements together, and as long as you are assigning it to a variable, you don't care what it is.

Why would you want to encourage the `button-a-123` pattern?

cahrens replied:
I still disagree that the IDs in their current form cause pain (or at least, I haven't heard of pain… is it documented somewhere?). As long as the ID is short, it is simple to search for it in the DOM. And dev tools do a nice job of linking things together. I think the risk of people creating non-unique IDs outweighs the potential benefit of a human-readable ID.

![image](https://user-images.githubusercontent.com/484484/208507399-da1f15a6-d41b-4201-91ed-d0eecd9addcd.png)

msilva-broad replied:
even the original IDContainer solution has intentional durability of the generated id between render cycles to avoid headaches of the id value mutating more often then expected (in between render cycles).  But the comments here have given me some great feedback in general.

I think I will simplify to @nawatts suggestion of having the hook simply return a durable unique ID, and have an optional argument at hook creation to allow for a prefix, which will default to 'element' and have a hypen added internally... matching #element-85 example in screen shot above.

msilva-broad committed:
IA-3762 useUniqueIdFn helper hook
- based on PR feedback, simplified hook to be useUniqueId and return the id directly instead of a callable function.
- useUniqueId takes optional prefix at hook creation time, will default to 'element' to match previous behavior.
- see updated jsdocs and unit tests.

msilva-broad committed:
Merge branch 'dev' into msilva/IA-3762_useUniqueIdFn

nawatts commented:
Just discovered this function already existed in react-utils:

https://github.com/DataBiosphere/terra-ui/blob/02959aec2697962cfe20aa33643c85ee53a14a88/src/libs/react-utils.js#L48-L50

cahrens commented:
@msilva-broad is out for the rest of the year, so I created https://broadworkbench.atlassian.net/browse/UIE-31 so we don't forget about this.
